File: apps/ios/BananaTape/BananaTape/InboundShareAdapters.swift
import Foundation

struct InboundImageCandidate: Equatable {
    let id: String
    let mimeType: ImageMimeType
    let originalFileName: String
    let sourceURL: URL
}

enum InboundShareStatus: Equatable {
    case idle
    case needsProjectSelection(InboundImageCandidate)
    case needsImportRole(project: MobileProjectRecord, candidate: InboundImageCandidate)
    case imported(ProjectImageAsset)
    case rejected(message: String)
}

enum InboundShareChoice: Equatable {
    case baseImage
    case referenceImage

    var importRole: ImportedImageRole {
        switch self {
        case .baseImage:
            return .baseImage
        case .referenceImage:
            return .referenceImage
        }
    }
}

struct InboundShareItem: Equatable {
    let mimeType: String
    let originalFileName: String
    let sourceURL: URL
}

struct InboundShareRejection: Error, Equatable {
    let message: String
}

struct InboundShareAdapter {
    static let unsupportedMessage = "Use a PNG or JPEG image."
    static let multiImageMessage = "Share one PNG or JPEG image at a time."
    static let chooseProjectMessage = "Choose or create a project before importing this image."
    static let chooseRoleMessage = "Choose whether to add this image as a base image or a reference."

    func candidate(from items: [InboundShareItem], id: String) -> Result<InboundImageCandidate, InboundShareRejection> {
        guard items.count == 1, let item = items.first else {
            return .failure(InboundShareRejection(message: Self.multiImageMessage))
        }
        guard let mimeType = ImageMimeType(rawValue: item.mimeType), mimeType == .png || mimeType == .jpeg else {
            return .failure(InboundShareRejection(message: Self.unsupportedMessage))
        }
        return .success(InboundImageCandidate(id: id, mimeType: mimeType, originalFileName: item.originalFileName, sourceURL: item.sourceURL))
    }
}

final class InboundShareModel {
    private(set) var status: InboundShareStatus = .idle
    private let storage: LocalProjectStorage
    private let adapter: InboundShareAdapter

    init(storage: LocalProjectStorage, adapter: InboundShareAdapter = InboundShareAdapter()) {
        self.storage = storage
        self.adapter = adapter
    }

    func receive(items: [InboundShareItem], activeProject: MobileProjectRecord?, candidateID: String) {
        switch adapter.candidate(from: items, id: candidateID) {
        case .success(let candidate):
            if let activeProject {
                status = .needsImportRole(project: activeProject, candidate: candidate)
            } else {
                status = .needsProjectSelection(candidate)
            }
        case .failure(let rejection):
            status = .rejected(message: rejection.message)
        }
    }

    func chooseProject(_ project: MobileProjectRecord) {
        guard case .needsProjectSelection(let candidate) = status else { return }
        status = .needsImportRole(project: project, candidate: candidate)
    }

    func importPending(as choice: InboundShareChoice) {
        guard case .needsImportRole(let project, let candidate) = status else { return }
        let result = storage.importProjectImage(ProjectImageImportRequest(
            projectID: project.id,
            assetID: candidate.id,
            role: choice.importRole,
            mimeType: candidate.mimeType,
            originalFileName: candidate.originalFileName,
            sourceURL: candidate.sourceURL
        ))
        switch result {
        case .success(let asset):
            status = .imported(asset)
        case .failure(let error):
            status = .rejected(message: error.userMessage)
        }
    }
}

File: apps/android/app/src/main/java/app/bananatape/mobile/adapters/InboundShareAdapters.kt
package app.bananatape.mobile.adapters

import android.content.Intent
import android.net.Uri
import java.nio.file.Path

data class InboundShareItem(
    val mimeType: String,
    val originalFileName: String,
    val sourcePath: Path,
)

data class InboundImageCandidate(
    val id: String,
    val mimeType: ImageMimeType,
    val originalFileName: String,
    val sourcePath: Path,
)

sealed interface InboundShareStatus {
    data object Idle : InboundShareStatus
    data class NeedsProjectSelection(val candidate: InboundImageCandidate) : InboundShareStatus
    data class NeedsImportRole(val project: MobileProjectRecord, val candidate: InboundImageCandidate) : InboundShareStatus
    data class Imported(val asset: ProjectImageAsset) : InboundShareStatus
    data class Rejected(val message: String) : InboundShareStatus
}

enum class InboundShareChoice(val importRole: ImportedImageRole) {
    BASE_IMAGE(ImportedImageRole.BASE_IMAGE),
    REFERENCE_IMAGE(ImportedImageRole.REFERENCE_IMAGE),
}

class InboundShareAdapter {
    fun candidate(items: List<InboundShareItem>, id: String): Result<InboundImageCandidate> {
        if (items.size != 1) return Result.failure(InboundShareRejected(MultiImageMessage))
        val item = items.single()
        val mimeType = when (item.mimeType) {
            ImageMimeType.PNG.value -> ImageMimeType.PNG
            ImageMimeType.JPEG.value -> ImageMimeType.JPEG
            else -> return Result.failure(InboundShareRejected(UnsupportedMessage))
        }
        return Result.success(InboundImageCandidate(id = id, mimeType = mimeType, originalFileName = item.originalFileName, sourcePath = item.sourcePath))
    }

    companion object {
        const val UnsupportedMessage = "Use a PNG or JPEG image."
        const val MultiImageMessage = "Share one PNG or JPEG image at a time."
        const val ChooseProjectMessage = "Choose or create a project before importing this image."
        const val ChooseRoleMessage = "Choose whether to add this image as a base image or a reference."
    }
}

class InboundShareRejected(val userMessage: String) : IllegalArgumentException(userMessage)

class InboundShareModel(
    private val importer: (ProjectImageImportRequest) -> AdapterResult<ProjectImageAsset>,
    private val adapter: InboundShareAdapter = InboundShareAdapter(),
) {
    var status: InboundShareStatus = InboundShareStatus.Idle
        private set

    fun receive(items: List<InboundShareItem>, activeProject: MobileProjectRecord?, candidateId: String) {
        status = adapter.candidate(items, candidateId).fold(
            onSuccess = { candidate ->
                if (activeProject == null) {
                    InboundShareStatus.NeedsProjectSelection(candidate)
                } else {
                    InboundShareStatus.NeedsImportRole(project = activeProject, candidate = candidate)
                }
            },
            onFailure = { error -> InboundShareStatus.Rejected(message = (error as InboundShareRejected).userMessage) },
        )
    }

    fun chooseProject(project: MobileProjectRecord) {
        val current = status
        if (current is InboundShareStatus.NeedsProjectSelection) {
            status = InboundShareStatus.NeedsImportRole(project = project, candidate = current.candidate)
        }
    }

    fun importPending(choice: InboundShareChoice) {
        val current = status
        if (current !is InboundShareStatus.NeedsImportRole) return
        status = when (val result = importer(ProjectImageImportRequest(current.project.id, current.candidate.id, choice.importRole, current.candidate.mimeType, current.candidate.originalFileName, current.candidate.sourcePath))) {
            is AdapterResult.Success -> InboundShareStatus.Imported(result.value)
            is AdapterResult.Failure -> InboundShareStatus.Rejected(result.error.userMessage)
        }
    }
}

class AndroidInboundIntentAdapter {
    fun deferredMessage(intent: Intent): String? {
        return when (intent.action) {
            Intent.ACTION_SEND_MULTIPLE -> InboundShareAdapter.MultiImageMessage
            Intent.ACTION_SEND, Intent.ACTION_VIEW -> null
            else -> InboundShareAdapter.UnsupportedMessage
        }
    }

    fun singleContentUri(intent: Intent): Uri? {
        if (intent.action == Intent.ACTION_SEND_MULTIPLE) return null
        return intent.getParcelableExtra(Intent.EXTRA_STREAM) ?: intent.data
    }
}
